storage/remote/read.go

```diff
@@ -95,9 +96,26 @@ func (q *querier) read(ctx context.Context, from, through model.Time, matchers m
 	}
 
 	removeLabels(res, added)
+	validateNamesAndValues(res)
+
 	return res, err
 }
 
+// validateNamesAndValues rejects invalid metric/label names returned from remote read.
+func validateNamesAndValues(res model.Matrix) {
+	for i, r := range res {
+		for name, value := range r.Metric {
+			if !model.IsValidMetricName(model.LabelValue(name)) || !utf8.ValidString(string(value)) {
```

Comment:
The rules for label names are stricter than for metric names, you'll need to check for valid label names.